msprime/demography.py

```diff
@@ -122,9 +122,36 @@ def insert_populations(self, tables):
         Insert population definitions for this demography into the specified
         set of tables.
         """
+        metadata_schema = tskit.MetadataSchema(
+            {
+                "codec": "json",
+                "type": "object",
+                "properties": {
+                    "name": {"type": "string"},
+                    "description": {"type": ["string", "null"]},
+                },
+                "required": ["name", "description"],
+                "additionalProperties": True,
+            }
+        )
         assert len(tables.populations) == 0
+        tables.populations.metadata_schema = metadata_schema
         for population in self.populations:
-            tables.populations.add_row(metadata=population.temporary_hack_for_metadata)
+            metadata = {
+                "name": population.name,
+                "description": population.description,
+            }
+            if population.extra_metadata is not None:
+                intersection = set(population.extra_metadata.keys()) & set(
+                    metadata.keys()
+                )
+                if len(intersection) > 0:
+                    warnings.warn(
+                        f"Overwriting standard metadata keys: {list(intersection)}; "
+                        "this may lead to compatibility problems."
```

Comment:
Is this warning right? Isn't it "existing" keys, not necessarily "standard" keys? Also, "compatibility" problems is pretty vague and alarming - I'd say either say concretely what the problems might be, or else just warn "overwriting existing keys: { }".

Reply:
Well, we do hope to standardise these (``name`` and ``description``) - and they are definitely standard metadata for msprime. I take your point about the warning, but it's hard to pinpoint what might happen downstream if you change the name of the population to something else. Probably nothing, but then why would you want to overwrite the name or description when you can set these as properties on the Population object in the first place. So - maybe this should be an error then, since you can always set the ``name`` and ``description`` via the Population class and it's probably a mistake if you're overwriting these.

@grahamgower, any thoughts here? I'm imagining stdpopsim as being one of the key users of this metadata API. Should this be a warning or an error?

Reply:
Probably it should be an error. I also can't see any reason to modify these, so it would most likely be done only accidentally.

Reply:
It might be because of my comment that it's a pain if a long-running script errors out after (say) 2 days or running, just before saving the final data files. I guess that's good reason to make these sorts of changes before you enter long-running loops, but that's not always the natural thing to do. 

I've definitely had problems with long running scripts failing at the very end due to things that were not vital to do, but which I would like to have been warned about. But I probably just have bad coding practices.

Reply:
Right, but I don't see how this would happen for an msprime simulation. We're not talking about general rules here, just what to do in this specific situtation.

Reply:
Ah, good point. Seems sensible to error out then.

Reply:
OK, I'll change this to an error and then I think we're good to merge.

Reply:
I agree - error in this case.